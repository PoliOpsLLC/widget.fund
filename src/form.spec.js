import { h, render } from 'preact';
import nock from 'nock';

import Form from './form';

const flushPromises = () => new Promise(resolve => setImmediate(resolve));

const packParams = params => {
    return Object.keys(params).map(key => {
        return `${encodeURIComponent(key)}=${encodeURIComponent(params[key])}`;
    }).join('&');
};

const simulateOn = (target, name, props = {}) => {
    const node = document.querySelector(target);
    const event = new Event(name, { bubbles: true });
    Object.assign(event, props);
    node.dispatchEvent(event);
};

describe('Form', () => {
    let result, scratch;

    const mount = async (overrides = {}) => {
        const props = {
            packParams,
            location_url: 'http://localhost/location',
            employer_url: 'http://localhost/employer',
            local_url: 'http://localhost/local',
            submit_url: 'http://localhost/submit',
            apiKey: 'API_KEY',
            token: 'TOKEN',
            ...overrides,
        };
        render(<Form ref={c => result = c} {...props} />, scratch);

        // one flush for each GET
        await flushPromises();
        await flushPromises();
        await flushPromises();
    };

    beforeAll(() => {
        global.Date.now = jest.fn(() => 1537832579326);
        scratch = document.createElement('div');
        document.body.appendChild(scratch);
    });

    beforeEach(() => {
        scratch.innerHTML = '';
        nock('http://localhost').get('/location').query(true)
            .reply(200, {
                results: [
                    { value: 'IA', display_name: 'Iowa' },
                    { value: 'LA', display_name: 'Louisiana' },
                ],
            });
        nock('http://localhost').get('/employer').query(true)
            .reply(200, {
                results: [
                    { address: { state: 'IA' }, locals: [1], name: 'Sample 3', pk: 3 },
                    { address: { state: 'LA' }, locals: [], name: 'Sample 4', pk: 4 },
                    { address: { state: 'LA' }, locals: [2], name: 'Sample 5', pk: 5 },
                ],
            });
        nock('http://localhost').get('/local').query(true)
            .reply(200, {
                results: [
                    { address: { state: 'LA' }, employers: [5], name: 'Sample 2', pk: 2 },
                    { address: { state: 'IA' }, employers: [], name: 'Sample 6', pk: 6 },
                    { address: { state: 'IA' }, employers: [3], name: 'Sample 1', pk: 1 },
                    { address: null, employers: [], name: 'Sample 7', pk: 7 },
                ],
            });
    });

    afterEach(() => {
        nock.cleanAll();
    });

    afterAll(() => {
        scratch.parentNode.removeChild(scratch);
        scratch = null;
    });

    it('loads all options on render', async done => {
        await mount();
        // wait for setState and its callback to finish at end of promise
        process.nextTick(() => {
            expect(result.base).toMatchSnapshot();
            expect(result.state).toMatchSnapshot();
            done();
        });
    });

    it('constrains employers based on state/local', async done => {
        await mount();
        // wait for setState and its callback to finish at end of promise
        process.nextTick(() => {
            simulateOn('input[name^="employer"]', 'focus');
            expect(document.querySelector('input[name^="employer"] ~ ul')).toMatchSnapshot();

            simulateOn('input[name^="location"]', 'focus');
            simulateOn('input[name^="location"] ~ ul li:last-child', 'mousedown', { button: 0 });
            simulateOn('input[name^="employer"]', 'focus');
            expect(document.querySelector('input[name^="employer"] ~ ul')).toMatchSnapshot();

            simulateOn('input[name^="local"]', 'focus');
            simulateOn('input[name^="local"] ~ ul li:first-child', 'mousedown', { button: 0 });
            simulateOn('input[name^="employer"]', 'focus');
            expect(document.querySelector('input[name^="employer"] ~ ul')).toMatchSnapshot();
            done();
        });
    });

    it('constrains locals based on state/employer', async done => {
        await mount();
        // wait for setState and its callback to finish at end of promise
        process.nextTick(() => {
            simulateOn('input[name^="local"]', 'focus');
            expect(document.querySelector('input[name^="local"] ~ ul')).toMatchSnapshot();

            simulateOn('input[name^="location"]', 'focus');
            simulateOn('input[name^="location"] ~ ul li:first-child', 'mousedown', { button: 0 });
            simulateOn('input[name^="local"]', 'focus');
            expect(document.querySelector('input[name^="local"] ~ ul')).toMatchSnapshot();

            simulateOn('input[name^="employer"]', 'focus');
            simulateOn('input[name^="employer"] ~ ul li:first-child', 'mousedown', { button: 0 });
            simulateOn('input[name^="local"]', 'focus');
            expect(document.querySelector('input[name^="local"] ~ ul')).toMatchSnapshot();
            done();
        });
    });

    it('redirects on submit with all field values', async done => {
        await mount();
        // wait for setState and its callback to finish at end of promise
        process.nextTick(() => {
            simulateOn('input[name^="location"]', 'focus');
            simulateOn('input[name^="location"] ~ ul li:first-child', 'mousedown', { button: 0 });
            simulateOn('input[name^="employer"]', 'focus');
            simulateOn('input[name^="employer"] ~ ul li:first-child', 'mousedown', { button: 0 });
            simulateOn('input[name^="local"]', 'focus');
            simulateOn('input[name^="local"] ~ ul li:first-child', 'mousedown', { button: 0 });

            global.parent.location.assign = jest.fn();
            simulateOn('button[type="submit"]', 'click');
            expect(global.parent.location.assign).toHaveBeenCalledWith(
                'http://localhost/submit?employer=3&local=1&location=IA&key=API_KEY&token=TOKEN'
            );
            done();
        });
    });
});
