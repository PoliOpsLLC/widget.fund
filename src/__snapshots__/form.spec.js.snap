// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form constrains employers based on state/local 1`] = `
<ul>
  <li
    aria-selected="false"
  >
    Sample 3
  </li>
  <li
    aria-selected="false"
  >
    Sample 4
  </li>
  <li
    aria-selected="false"
  >
    Sample 5
  </li>
</ul>
`;

exports[`Form constrains employers based on state/local 2`] = `
<ul>
  <li
    aria-selected="false"
  >
    Sample 4
  </li>
  <li
    aria-selected="false"
  >
    Sample 5
  </li>
</ul>
`;

exports[`Form constrains employers based on state/local 3`] = `
<ul>
  <li
    aria-selected="false"
  >
    Sample 5
  </li>
</ul>
`;

exports[`Form constrains locals based on state/employer 1`] = `
<ul>
  <li
    aria-selected="false"
  >
    Sample 1
  </li>
  <li
    aria-selected="false"
  >
    Sample 2
  </li>
  <li
    aria-selected="false"
  >
    Sample 6
  </li>
  <li
    aria-selected="false"
  >
    Sample 7
  </li>
</ul>
`;

exports[`Form constrains locals based on state/employer 2`] = `
<ul>
  <li
    aria-selected="false"
  >
    Sample 1
  </li>
  <li
    aria-selected="false"
  >
    Sample 6
  </li>
  <li
    aria-selected="false"
  >
    Sample 7
  </li>
</ul>
`;

exports[`Form constrains locals based on state/employer 3`] = `
<ul>
  <li
    aria-selected="false"
  >
    Sample 1
  </li>
</ul>
`;

exports[`Form loads all options on render 1`] = `
<div
  class="content"
>
  <h3>
    This is the widget form
  </h3>
  <p>
    Each one of the fields below help narrow down which organization the member belongs to.
  </p>
  <form>
    <div
      class="typeahead"
    >
      <label
        for="location-1537832579326"
      >
        State of Employment
      </label>
      <div
        class="awesomplete"
      >
        <input
          aria-autocomplete="list"
          autocomplete="off"
          name="location-1537832579326"
        />
        <ul
          hidden=""
        />
        <span
          aria-live="assertive"
          aria-relevant="additions"
          class="visually-hidden"
          role="status"
        />
      </div>
    </div>
    <div
      class="typeahead"
    >
      <label
        for="employer-1537832579326"
      >
        Employer
      </label>
      <div
        class="awesomplete"
      >
        <input
          aria-autocomplete="list"
          autocomplete="off"
          name="employer-1537832579326"
        />
        <ul
          hidden=""
        />
        <span
          aria-live="assertive"
          aria-relevant="additions"
          class="visually-hidden"
          role="status"
        />
      </div>
    </div>
    <div
      class="typeahead"
    >
      <label
        for="local-1537832579326"
      >
        Affiliate/Local
      </label>
      <div
        class="awesomplete"
      >
        <input
          aria-autocomplete="list"
          autocomplete="off"
          name="local-1537832579326"
        />
        <ul
          hidden=""
        />
        <span
          aria-live="assertive"
          aria-relevant="additions"
          class="visually-hidden"
          role="status"
        />
      </div>
    </div>
    <p>
      Submitting will take what has been entered and redirect the member to the matched organization's forms to complete their sign up process.
    </p>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`Form loads all options on render 2`] = `
Object {
  "employer": "",
  "employerOptions": Array [
    Object {
      "address": Object {
        "state": "IA",
      },
      "label": "Sample 3",
      "locals": Array [
        1,
      ],
      "name": "Sample 3",
      "pk": 3,
      "value": 3,
    },
    Object {
      "address": Object {
        "state": "LA",
      },
      "label": "Sample 4",
      "locals": Array [],
      "name": "Sample 4",
      "pk": 4,
      "value": 4,
    },
    Object {
      "address": Object {
        "state": "LA",
      },
      "label": "Sample 5",
      "locals": Array [
        2,
      ],
      "name": "Sample 5",
      "pk": 5,
      "value": 5,
    },
  ],
  "local": "",
  "localOptions": Array [
    Object {
      "address": Object {
        "state": "LA",
      },
      "employers": Array [
        5,
      ],
      "label": "Sample 2",
      "name": "Sample 2",
      "pk": 2,
      "value": 2,
    },
    Object {
      "address": Object {
        "state": "IA",
      },
      "employers": Array [],
      "label": "Sample 6",
      "name": "Sample 6",
      "pk": 6,
      "value": 6,
    },
    Object {
      "address": Object {
        "state": "IA",
      },
      "employers": Array [
        3,
      ],
      "label": "Sample 1",
      "name": "Sample 1",
      "pk": 1,
      "value": 1,
    },
    Object {
      "address": null,
      "employers": Array [],
      "label": "Sample 7",
      "name": "Sample 7",
      "pk": 7,
      "value": 7,
    },
  ],
  "location": "",
  "locationOptions": Array [
    Object {
      "display_name": "Iowa",
      "label": "Iowa",
      "value": "IA",
    },
    Object {
      "display_name": "Louisiana",
      "label": "Louisiana",
      "value": "LA",
    },
  ],
}
`;
